File: src/main/scala/com/vc/hard/L336.scala
package com.vc.hard

object L336 {
  def palindromePairs(words: Array[String]): List[List[Int]] = {
    import scala.collection.mutable
    val map = new mutable.HashMap[String, Int]()
    words.indices.foreach(i => {
      map.put(words(i), i)
    })
    val res = new mutable.ListBuffer[List[Int]]()
    def isPalindrome(str: String): Boolean = {
      var lo = 0
      var hi = str.length - 1
      while(lo <= hi) {
        if(str(lo) != str(hi)) return false
        lo += 1
        hi -= 1
      }
      true
    }
    words.indices.foreach(i => {
      (0 to words(i).length).foreach(j => {
        val str1 = words(i).substring(0, j)
        val str2 = words(i).substring(j)

        if(isPalindrome(str1)) {
          val str2Reverse = str2.reverse
          if(map.contains(str2Reverse)) {
            val k = map(str2Reverse)
            if(i != k) {
              val list = List(k, i)
              res += list
            }
          }
        }

        if(isPalindrome(str2) && str2.length != 0) {
          val str1Reverse = str1.reverse
          if(map.contains(str1Reverse)) {
            val k = map(str1Reverse)
            if(i != k) {
              val list = List(i, k)
              res += list
            }
          }
        }
      })
    })
    res.toList
  }
}

File: src/main/java/com/vc/hard/L336.java
package com.vc.hard;

import java.util.*;

class L336 {
    public List<List<Integer>> palindromePairs(String[] words) {
        HashMap<String, Integer> wordIndex = new HashMap<>();
        for(int i = 0; i< words.length; i++) {
            wordIndex.put(words[i], i);
        }
        //System.out.println(wordIndex);
        List<List<Integer>> res = new ArrayList<List<Integer>>();

        for(int i = 0; i < words.length; i++) {
            String word = words[i];
            for(int j = 0; j <= word.length(); j++) {
                String w1 = word.substring(0, j);
                String w2 = word.substring(j, word.length());

                if(isPalindrome(w1)) {
                    String w2Reverse = reverse(w2);
                    if(wordIndex.containsKey(w2Reverse)) {
                        int k = wordIndex.get(w2Reverse);
                        if(i != k) res.add(Arrays.asList(k, i));
                    }
                }

                if(isPalindrome(w2) && w2.length() != 0) {
                    String w1Reverse = reverse(w1);
                    if(wordIndex.containsKey(w1Reverse)) {
                        int k = wordIndex.get(w1Reverse);
                        if(i != k) res.add(Arrays.asList(i, k));
                    }
                }
            }
        }
        return res;
    }


    private String reverse(String str){
        StringBuilder s = new StringBuilder(str);
        return s.reverse().toString();
    }

    private boolean isPalindrome(String str) {
        int left = 0;
        int right = str.length() - 1;
        while(left < right) {
            if(str.charAt(left) != str.charAt(right)) return false;
            left++;
            right--;
        }
        return true;
    }
}
